doc/architecture/decisions/0015-container-api.md

```diff
@@ -0,0 +1,103 @@
+
+A Holochain app cannot be used directly. Holochain is built as a library that exposes an interface for
+creating Holochain instances and calling their zome functions (see [core_api](/core_api/src/lib.rs)).
+This *core_api* requires its client to provide a context, which holds the representation of the agent
+(name and keys), a logger and persistence and will also include a networking proxy in the future.
+
+While it is possible to use this library directly in a monolithic app where the UI is tightly
+coupled to the app and everything linked to the Holochain library into the same executable, we regard this
+as a special case, since this does not allow for app composability (which we think is a crucial
+concept for the Holochain app ecosystem). 
+
+Instead, in the context of the Rust iteration of Holochain, we have been following the notion of
+providing a *Container* as a relevant concept (and implemention) for the deployment of Holochain apps:
+we provide a Container (i.e. [HoloSqape](https://github.com/holochain/holosqape)) for each supported platform (Linux, MacOS, Windows, iOS, Android)
+that gets installed on a host machine. Holochain apps get installed into the container.
+The Container:  
+ * manages installed hApps (Holochain Apps),
+ * manages agent identities (keys) ,
+ * should also enable hApps to call functions of other hApps - what we call *bridging*,
+ * has to implement access permissions to installed hApps.
+
+So far, the interface our Container implementation provides was growing organically
+in [container.h](https://github.com/holochain/holosqape/blob/master/bindings/container.h).
+
+With upcoming alternative container implementations (based on [Holochain-nodejs](https://github.com/holochain/holochain-nodejs)
+or a custom one for HoloPorts) we should drive the process of building this Container API
+consciously and with more coherence and visibility amongst our separate dev teams.
+
+We need a protocol for communication with a Holochain container and a specification of what upcoming
+containers have to implement, so that apps and UIs can be build against a standardised
+
+## Decision
+
+We establish the **Container as a fundamental module of the Holochain architecture/stack**
+by specifying its **API**, that can be assumed by UI components, Holochain apps (i.e. zome
+functions in the case of bridging)
+and remote processes alike to be implemented by the context a Holochain app is executed in.
+
+
+Fundamental to this API is user roles and permissions.
+Users will be able to use different subsets of the Container's API depending on their specific permissions. 
+
+### Generic API
+
+Every user (of the Container API, i.e. QML root widgets, admin UI in the case of HoloSqape, 
+and network services built on top of the Container) 
+will have these functions available, though functions can return with a permission denied
+error in case an app or capability was given that the current user is not allowed to use.
```

Comment:
Just for clarity, are capabilities issued to dpki umbrellas, and "`user`" in this case is one authorized device key under that umbrella?

Reply:
Hm, in this case, `user` means user of the container API, which might be:
* some form of local UI (QML root UI in the case of HoloSqape, but could be the Holo hosting app for example)
* another app
* something that is calling in via a network connection, if the contianer is embedded in a service like application